ramips: get MAC from the encrypted partition (WG4хх223) - fix issue #10062

csharper2005 opened the pull request:
~~**This depends on** https://github.com/openwrt/openwrt/pull/10222~~ (merged)

This commit adds decryption of the Arcadyan WG4xx223 configuration
partition (board_data) and get MAC address from it. After this change
the hack with saving MAC addressees to u-boot-env before installation
of OpenWrt is no longer necessary.

Affected routers
----------------
- Beeline Smartbox Flash (Arcadyan WG443223)
- MTS WG430223 (Arcadyan WG430223)

Related issue
-------------
Arcadyan WG4хх223 routers: getting MACs from the encrypted partitions)
Link:
https://github.com/openwrt/openwrt/issues/10062

Related PR
----------
uencrypt: add package to decrypt WG4хх223 config
Link:
https://github.com/openwrt/openwrt/pull/10222

csharper2005 committed:
ramips: get MAC addr from the encrypted partition (WG4хх223)

This commit resolves #10062. Adds decryption of the Arcadyan WG4xx223
configuration partition (board_data)to get base MAC address from it.
As a result, after this change the hack with saving MAC addressees to
u-boot-env before installation of OpenWrt is no longer necessary.

This is necessary for the following devices:
- Beeline Smartbox Flash (Arcadyan WG443223)
- MTS WG430223 (Arcadyan WG430223)

Example:
+----------------+-------------------+------------------------+
|                |    MTS WG430223   | Beeline Smartbox Flash |
+----------------+-------------------+------------------------+
| base mac (mtd) | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:06      |
| label          | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:09      |
| LAN            | A4:xx:xx:51:xx:F6 | 30:xx:xx:51:xx:09      |
| WAN            | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:06      |
| WLAN_2g        | A4:xx:xx:51:xx:F5 | 30:xx:xx:51:xx:07      |
| WLAN_5g        | A6:xx:xx:21:xx:F5 | 32:xx:xx:41:xx:07      |
+----------------+-------------------+------------------------+

Collected statistic shows that the 2-4th bits of the 7th byte of the
WLAN_5g MAC are the constant (see #10062 for more details):
- Beeline Smartbox Flash - 100
- MTS WG430223 - 010

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 commented:
Do we need a separate function to change the whole byte? Or 4 calls of `macaddr_setbit / macaddr_unsetbit` for each device is ok?

cotequeiroz commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -79,6 +79,26 @@ mtd_get_mac_ascii() {
+
+	# config decryption and getting mac
+	mac_dirty=$(dd if="$part" bs=1 count=$((0x170)) skip=$((0x100)) 2>/dev/null | \
+		uencrypt -d -n -k 2A4B303D7644395C3B2B7053553C5200 \
```
I suggest to check if uencrypt is available; use it if found.  If not, check for the openssl util, which can be used instead.  I'd give uencrypt the preference, since it should run a little faster, using less memory.

csharper2005 replied:
@cotequeiroz nice idea. Done.

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/c83d8cc4a2edce3e89011e54abbe859537fc6b10 - add new `macaddr_setbyte `function

csharper2005 committed:
base-files: add mtd_get_mac_encrypted_arcadyan function

Some Arcadyan devices (e.g. MTS WG430223) keep their config in encrypted
mtd. This adds mtd_get_mac_encrypted_arcadyan() function to get the MAC
address from the encrypted partition. Function uses uencrypt utility for
decryption (and openssl if the uencrypt wasn't found).

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 commented:
@Ansuel 

> * Use Fixes tag for the issue
> * Drop the related PR thing
> * Rework the affected routers thing

Done. I hope :)

> Also can you provide an example for the use of the set mac bytes? I'm a bit confused so I'm asking about it.
> 

Example:
```
+----------------+-------------------+------------------------+
|                |    MTS WG430223   | Beeline Smartbox Flash |
+----------------+-------------------+------------------------+
| base mac (mtd) | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:06      |
| label          | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:09      |
| LAN            | A4:xx:xx:51:xx:F6 | 30:xx:xx:51:xx:09      |
| WAN            | A4:xx:xx:51:xx:F4 | 30:xx:xx:51:xx:06      |
| WLAN_2g        | A4:xx:xx:51:xx:F5 | 30:xx:xx:51:xx:07      |
| WLAN_5g        | A6:xx:xx:21:xx:F5 | 32:xx:xx:41:xx:07      |
+----------------+-------------------+------------------------+
```
An important feature of these routers... The 7th byte of the WLAN_5g MAC is a constant:
MTS WG430223 = 2
Beeline Smartbox Flash = 4

This is confirmed by statistic (I have 5-6 examples for each device model). 

So, would be nice to have a function to change the whole byte (`macaddr_setbyte`).

This changes the 7th byte of the `hw_mac_addr` to 4:
`hw_mac_addr=$(macaddr_setbyte $hw_mac_addr 7 4)`
A6:AA:AA:**5**1:AA:F5 - before 
A6:AA:AA:**4**1:AA:F5 - after

> Also this needs a rebase :D

Done

csharper2005 commented:
@Ansuel, @SopaDeMacaco-UmaDelicia found a Beeline Smartbox Flash device that brakes the previous theory (7th byte is a constant). So, I updated the issue record and calculation algorithm in this PR according to this finding.

csharper2005 commented:
Close due to https://github.com/openwrt/openwrt/issues/10062#issuecomment-1204254537. Current calculation algorithm is still unknown.

Ansuel commented:
@csharper2005 ok how to proceed with this? 
uencrypt should also be backported right?

csharper2005 commented:
> @csharper2005 ok how to proceed with this? uencrypt should also be backported right?

@Ansuel yeah, for 22.03 we need this + uencrypt. For master we need this PR only.

> Is it ready so I can give a final review?

Yes, from the technical side it works. I have tested it myself on both devices, with both utils (uencrypt and openssl) and also have positive feedback from users (my 22.03-rc community builds).

Ansuel commented:
Merged in master with
a6b0d08060 ramips: get MAC addr from the encrypted partition (WG4хх223)
12c971bc26 base-files: add mtd_get_mac_encrypted_arcadyan function

Backported to openwrt-22.03 with
ecd8f7e294 ramips: get MAC addr from the encrypted partition (WG4хх223)
290ace2fe6 base-files: add mtd_get_mac_encrypted_arcadyan function
d94a28f7d2 uencrypt: add package to decrypt WG4хх223 config

@csharper2005 can you check if the backport applied correctly?

csharper2005 commented:
> Сan you check if the backport applied correctly?

@Ansuel, of course. I'll build the image and give you a feedback. Thanks a lot!

csharper2005 commented:
@Ansuel. All as expected on OpenWrt 22.03-SNAPSHOT, r19641-ecd8f7e294. Works fine.

Beeline SmartBox Flash:
```
root@OpenWrt:~# ifconfig | grep HWaddr
br-lan    Link encap:Ethernet  HWaddr EC:xx:xx:7F:xx:0B
eth0      Link encap:Ethernet  HWaddr 00:0C:43:28:80:A0
lan1      Link encap:Ethernet  HWaddr EC:xx:xx:7F:xx:0B
lan2      Link encap:Ethernet  HWaddr EC:xx:xx:7F:xx:0B
wan       Link encap:Ethernet  HWaddr EC:xx:xx:7F:xx:08
wlan0     Link encap:Ethernet  HWaddr EC:xx:xx:7F:xx:09
wlan1     Link encap:Ethernet  HWaddr EE:xx:xx:4F:xx:09
```

MTS WG430223:
```
root@OpenWrt:~# ifconfig | grep HWaddr
br-lan    Link encap:Ethernet  HWaddr A4:xx:xx:51:xx:F6
eth0      Link encap:Ethernet  HWaddr 00:0C:43:28:80:A0
lan1      Link encap:Ethernet  HWaddr A4:xx:xx:51:xx:F6
lan2      Link encap:Ethernet  HWaddr A4:xx:xx:51:xx:F6
wan       Link encap:Ethernet  HWaddr A4:xx:xx:51:xx:F4
wlan0     Link encap:Ethernet  HWaddr A4:xx:xx:51:xx:F5
wlan1     Link encap:Ethernet  HWaddr A6:xx:xx:21:xx:F5
```

csharper2005 commented:
Master (OpenWrt SNAPSHOT, r20357-8c04a5c456) is also works fine. @Ansuel can we close this PR?
